rasterio/__init__.py

```diff
@@ -190,46 +189,60 @@ def open(fp, mode='r', driver=None, width=None, height=None, count=None,
             "Blacklisted: file cannot be opened by "
             "driver '{0}' in '{1}' mode".format(driver, mode))
 
-    # Special case for file object argument.
+    # If the fp argument is a file-like object and can be adapted by
+    # rasterio's FilePath we do so. Otherwise, we use a MemoryFile to
+    # hold fp's contents and store that in an ExitStack attached to the
+    # dataset object that we will return. When a dataset's close method
+    # is called, this ExitStack will be unwound and the MemoryFile's
+    # storage will be cleaned up.
     if mode == 'r' and hasattr(fp, 'read'):
-
-        @contextmanager
-        def fp_reader(fp):
-            if have_vsi_plugin:
-                vsi_file = FilePath(fp)
-            else:
-                vsi_file = MemoryFile(fp.read())
-            dataset = vsi_file.open(driver=driver, sharing=sharing)
-            try:
-                yield dataset
-            finally:
-                dataset.close()
-                vsi_file.close()
-
-        return fp_reader(fp)
+        if have_vsi_plugin:
+            return FilePath(fp).open(driver=driver, sharing=sharing)
+        else:
+            memfile = MemoryFile(fp.read())
+            dataset = memfile.open(driver=driver, sharing=sharing)
+            ctxstack = ExitStack()
+            ctxstack.enter_context(env_ctx_if_needed())
+            ctxstack.enter_context(memfile)
+            dataset._env = ctxstack
+            return dataset
```

Comment:
@sgillies I'm not sure to see where `ctxstack.close()` gets called!

Reply:
At https://github.com/rasterio/rasterio/pull/2371/files#diff-9729b3413c39ccb6eea16562e513703922f21d4d94a3e27980926dd01a0c5068R438 the equivalent happens, but yes, it's probably more clear to use `self._env.ctx.close()` :+1: